internal/rukpak/source/containers_image.go

```diff
@@ -0,0 +1,274 @@
+package source
+
+import (
+	"archive/tar"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/containerd/containerd/archive"
+	"github.com/containers/image/v5/copy"
+	"github.com/containers/image/v5/docker"
+	"github.com/containers/image/v5/docker/reference"
+	"github.com/containers/image/v5/manifest"
+	"github.com/containers/image/v5/oci/layout"
+	"github.com/containers/image/v5/pkg/blobinfocache/none"
+	"github.com/containers/image/v5/pkg/compression"
+	"github.com/containers/image/v5/signature"
+	"github.com/containers/image/v5/types"
+	"github.com/opencontainers/go-digest"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type ContainersImageRegistry struct {
+	BaseCachePath string
+	SourceContext *types.SystemContext
+}
+
+func (i *ContainersImageRegistry) Unpack(ctx context.Context, bundle *BundleSource) (*Result, error) {
+	l := log.FromContext(ctx)
+
+	if bundle.Type != SourceTypeImage {
+		panic(fmt.Sprintf("programmer error: source type %q is unable to handle specified bundle source type %q", SourceTypeImage, bundle.Type))
+	}
+
+	if bundle.Image == nil {
+		return nil, NewUnrecoverable(fmt.Errorf("error parsing bundle, bundle %s has a nil image source", bundle.Name))
+	}
+
+	//////////////////////////////////////////////////////
+	//
+	// Resolve a canonical reference for the image.
+	//
+	//////////////////////////////////////////////////////
+	imgRef, err := reference.ParseNamed(bundle.Image.Ref)
+	if err != nil {
+		return nil, NewUnrecoverable(fmt.Errorf("error parsing image reference %q: %w", bundle.Image.Ref, err))
+	}
+
+	canonicalRef, err := resolveCanonicalRef(ctx, imgRef, i.SourceContext)
+	if err != nil {
+		return nil, fmt.Errorf("error resolving canonical reference: %w", err)
+	}
+
+	//////////////////////////////////////////////////////
+	//
+	// Check if the image is already unpacked. If it is,
+	// return the unpacked directory.
+	//
+	//////////////////////////////////////////////////////
+	unpackPath := i.unpackPath(bundle.Name, canonicalRef.Digest())
+	if unpackStat, err := os.Stat(unpackPath); err == nil {
+		if !unpackStat.IsDir() {
+			return nil, fmt.Errorf("unexpected file at unpack path %q: expected a directory", unpackPath)
+		}
+		l.Info("image already unpacked", "ref", imgRef.String(), "digest", canonicalRef.Digest().String())
+		return successResult(bundle.Name, unpackPath, canonicalRef), nil
+	}
+	//if err := os.MkdirAll(i.bundlePath(bundle.Name), 0755); err != nil {
+	//	return nil, fmt.Errorf("error creating unpack directory: %w", err)
+	//}
```

Comment:
Remove this?

Reply:
https://github.com/operator-framework/operator-controller/compare/df384cd6cb07fbd53f8d037b94d24a96a60c50b9..f21e3e5003b70b5b14eb73fc3c5170dd4f147425